File: java/falsesecurity.java
import java.util.ArrayList;
import java.util.Scanner;

public class falsesecurity {
	static final char[] characters = new char[] { 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N',
			'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '_', '.', ',', '?' };
	static final String[] codes = new String[] { ".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---",
			"-.-", ".-..", "--", "-.", "---", ".--.", "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", "-.--",
			"--..", "..--", "---.", ".-.-", "----" };

	static int getIndex(char[] array, char value) {
		for (int i = 0; i < array.length; i++)
			if (array[i] == value)
				return i;
		return -1;
	}

	static int getIndex(String[] array, String value) {
		for (int i = 0; i < array.length; i++)
			if (array[i].equals(value))
				return i;
		return -1;
	}

	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		while (scan.hasNextLine()) {
			String input = scan.nextLine();
			StringBuilder morse = new StringBuilder();
			ArrayList<Integer> lengths = new ArrayList<>();
			int index = 0;
			for (char ch : input.toCharArray()) {
				index = getIndex(characters, ch);
				morse.append(codes[index]);
				lengths.add(codes[index].length());
			}
			index = 0;
			StringBuilder msg = new StringBuilder();
			for (int i = lengths.size() - 1; i >= 0; i--)
				msg.append(characters[getIndex(codes, morse.substring(index, index += lengths.get(i)))]);
			System.out.println(msg);
		}
		scan.close();
	}
}

File: ts/src/falsesecurity.ts
import readline = require('readline')

function decodeMsg (msg: string): string {
  const morse: string[] = []
  const lengths: number[] = []

  msg.split('').forEach((c) => {
    const index = chars.indexOf(c)
    morse.push(codes[index] ?? '')
    lengths.push(codes[index]?.length ?? 0)
  })

  const morseStr = morse.join('')
  let morseOffset = 0

  return lengths
    .reverse()
    .map((l) => {
      const index = codes.indexOf(morseStr.substr(morseOffset, l))
      morseOffset += l
      return chars[index]
    })
    .join('')
}

const chars: string[] = [
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z',
  '_',
  '.',
  ',',
  '?'
]

const codes: string[] = [
  '.-',
  '-...',
  '-.-.',
  '-..',
  '.',
  '..-.',
  '--.',
  '....',
  '..',
  '.---',
  '-.-',
  '.-..',
  '--',
  '-.',
  '---',
  '.--.',
  '--.-',
  '.-.',
  '...',
  '-',
  '..-',
  '...-',
  '.--',
  '-..-',
  '-.--',
  '--..',
  '..--',
  '---.',
  '.-.-',
  '----'
]

readline
  .createInterface({
    input: process.stdin,
    output: process.stdout
  })
  .on('line', (line: string) => {
    const msg = decodeMsg(line)
    console.log(msg)
  })
  .on('close', () => {
    process.exit(0)
  })
